genie-agent/src/main/java/com/netflix/genie/agent/execution/process/impl/JobProcessManagerImpl.java

```diff
@@ -200,14 +196,19 @@ public JobProcessResult waitFor() throws InterruptedException {
             timeoutThreadFuture.cancel(true);
         }
 
-        // TODO: This doesn't seem quite right to me and is confirmed by an existing test. If the job completes
-        //       successfully but then the system calls kill before this method then the final job status will be
-        //       killed instead of successful. We should discuss this - TJG 9/9/2019
-        if (this.killed.get()) {
-            final KillService.KillSource source = this.killSource.get() != null
-                ? this.killSource.get()
-                : KillService.KillSource.API_KILL_REQUEST;
+        // Check exit code first to see if the job finishes successfully and returns SUCCEEDED as status,
+        // even the job gets a KILL request.
+        if (process != null && exitCode == SUCCESS_EXIT_CODE) {
```

Comment:
I wonder if we need to synchronize on this kind of check. I know above it dereferences the atomic get but by here we're out of the lock